sparse/core.py

```diff
@@ -603,55 +613,238 @@ def elemwise_binary(self, func, other, *args, **kwargs):
+        dimension doesn't need broadcasting, and False if it needs broadcasting.
+        """
+        params = [None if l1 is None else l1 == l2 for l1, l2 in zip_longest(shape[::-1], broadcast_shape[::-1],
+                                                                             fillvalue=None)][::-1]
+
+        return params
+
+    @staticmethod
+    def _get_reduced_coords(coords, params):
+        reduced_params = [bool(param) for param in params]
+
+        return coords[reduced_params]
+
+    @staticmethod
+    def _get_expanded_coords_data(coords, data, params, broadcast_shape):
+        times_repeated = 1
+        coords_list = []
+        dim = 0
+        names = coords.dtype.names
+        data_len = len(data)
+        for param, l in zip(params, broadcast_shape):
+            if param:
+                coords_list.append(np.repeat(coords[names[dim]], times_repeated))
+            else:
+                data = np.repeat(data, l)
+                coords_list = [np.repeat(coord, l) for coord in coords_list]
+                coords_list.append(np.tile(np.arange(l), times_repeated * data_len))
+                times_repeated *= l
+
+            if param is not None:
+                dim += 1
 
-        # record array to ND array
-        coords = np.asarray(coords.view(coords.dtype[0]).reshape(len(coords), self.ndim)).T
+        return coords_list, data
+
+    def broadcast_to(self, shape):
+        result_shape = self._get_broadcast_shape(self.shape, shape, is_result=True)
+        params = self._get_broadcast_parameters(self.shape, result_shape)
+        self_coords = np.rec.fromarrays(self.coords)
+        coords, data = self._get_expanded_coords_data(self_coords, self.data, params, result_shape)
 
-        return COO(coords, data, shape=self.shape, has_duplicates=False)
+        return COO(coords, data, shape=result_shape, has_duplicates=self.has_duplicates)
+
+    @staticmethod
+    def _get_matching_coords(coords1, coords2, shape1, shape2):
+        result_shape = COO._get_broadcast_shape(shape1, shape2)
+        params1 = COO._get_broadcast_parameters(shape1, result_shape)
+        params2 = COO._get_broadcast_parameters(shape2, result_shape)
+
+        coords_list = []
+        dim1 = 0
+        names1 = coords1.dtype.names
+
+        dim2 = 0
+        names2 = coords2.dtype.names
+
+        for p1, p2 in zip(params1, params2):
+            if p1:
+                coords_list.append(coords1[names1[dim1]])
+            else:
+                coords_list.append(coords2[names2[dim2]])
+
+            if p1 is not None:
+                dim1 += 1
+
+            if p2 is not None:
+                dim2 += 1
+
+        return coords_list
+
+    @staticmethod
+    def _get_broadcast_coords_data(coords, matched_coords, data, params, broadcast_shape):
+        full_coords, full_data = COO._get_expanded_coords_data(coords, data, params, broadcast_shape)
+        full_coords = np.rec.fromarrays(full_coords)
```

Comment:
Why `np.rec.from_arrays` here?  Is this a record array?  If so, why?

Reply:
It turns out that `matched_coords` and `full_coords` aren't necessarily sorted on input to this function. To perform the sorting, I convert them to `recarray`s, because we need to perform matching.